### ISSUE TITLE
Connecting to a signal on child node instantiated in code doesn't work the same way as scene-instantiated children does

### ISSUE THREAD

**Issue by nightblade9** [godotengine/godot#83582]
### Godot version

4.1.2 official latest

### System information

Manjaro, latest, Kernel 5.5 I think

### Issue description

Spent a few hours tearing my hair out to figure this out. I have a `main.tscn` scene which contains a `monster_container` scene. The monster container is just a regular control with a script file. Under it, I have a single instance of a `monster` scene:

```
o MainScene
|__o MonsterContainer (Control with script)
     |__o Monster.tscn
```

In terms of code;
```gdscript

# main.gd
const Monster = preload("res://.../monster.tscn")

@onready var _monster_container = $...

func _ready():
  _monster_container.add_child(Monster.instantiate())

# monster_container.gd
func _ready():
  for monster in self.get_children():
    monster.died.connect(func(); print("IT WORKS!!!!!")

# monster.gd
 self.died.emit()
```

TL;DR I the lambda for `monster.died` triggers for the node that was already pre-existing in the tscn, but not the one created in code.

In terms of behaviour: I can see the signal being connected with breakpoints, verify that there's a method there (connecting twice throws an error), see it being emitted, but it's never received.

I debugged this code (I won't bore you with the details) and nailed it down to this issue: `monster_container._ready` was called before `monster._ready`. This might be obvious to you, but it wasn't to me.

moving my `monster_container` code into a function that's called after `_ready` executes for each `monster` instance, solved my dilemma.

Personally, I would like to see an error of some sort in this scenario. Ideally, Godot should tell me that I can't connect to `monster.died` because `monster` hasn't had `_ready` called on it yet.

### Steps to reproduce

See code sample above.

### Minimal reproduction project

[issue83582.zip](https://github.com/godotengine/godot/files/13044860/issue83582.zip)

**Comment by AThousandShips** [comment]
> Possibly available on request, if you request it.

Yes do provide a minimal reproduction project please, as this is not trivial to reproduce from your instructions

**Comment by nightblade9** [comment]
> Yes do provide a minimal reproduction project please, as this is not trivial to reproduce from your instructions

I made a new MVP from scratch, and ... it worked. That makes me suspicious that this is probably something wrong in my code. I tried to eviscerate my own project into an MVP, but it crashes on run.

I'll look into this further and see if I can figure out what's going on. If it looks like a bug after I get the MVP cut down, I'll let you know.

**Comment by nightblade9** [comment]
Updated my issue and got it to reproduce with a very minimal new project that I created from scratch (RIP the bones of my gutted project).

In case it wasn't clear, here's the problem:

- MonsterContainer has a single child: an instance of `Monster`
- I then add two new instances via code, by calling `_monster_container.add_child(Monster.instantiate())`
- I then kill all the monsters by clicking the button three times in the repro script
- The `died` signal emits fine for the monster that was already in the scene; not so for the two I added via code.

Ergo, it does look like a Godot bug to me, but maybe this is working as designed. (I only ask for an error message, not necessarily a functionality change, if this is the intended behaviour.)

**Comment by kleonc** [comment]
@nightblade9 Everything in the provided MRP works correctly / as intended. The order in which `_ready` methods are called for the nodes in the scene tree is well defined and documented. Hence it's expected that the monsters you're adding in `Node2D._ready` are added after the `Node2D/MonsterContainer._ready` has executed (hence you're not connecting any of their signals). This a design / architectural problem on your side, Godot works here as documented. Hence closing as not a bug.

Also I don't see any reason why any error message would be needed here, or what such error message should state. Everything works as intended, there's no error the user needs to be informed about.

---

And here's a potential solution for your problem: instead of connecting signals only of the children existing during `_ready`:
```gdscript
extends Control

func _ready():
	for monster in get_children():
		monster.died.connect(_on_monster_died)

func _on_monster_died(who) -> void:
	print("I got the signal that %s died." % who)
```
you could connect each child's signal when such child is entering the tree, by connecting to the `MonsterManager`'s [`Node.child_entered_tree`](https://docs.godotengine.org/en/4.1/classes/class_node.html#class-node-signal-child-entered-tree) signal:
```gdscript
extends Control

func _on_child_entered_tree(node: Node) -> void:
	if node.has_signal("died"):
		node.died.connect(_on_monster_died)

func _on_monster_died(who) -> void:
	print("I got the signal that %s died." % who)
```

**Comment by nightblade9** [comment]
> (hence you're not connecting any of their signals)

I'm not sure I understand. If you add a breakpoint to the line where we connect to `monster.died` in the parent container, you see that a) the node we're connecting to is defined and non-null, and b) the signal gets connected three times.

I don't see how that could be the case if I'm adding the nodes later, after `_ready`.

**Comment by nightblade9** [comment]
Hmm, interesting. You're correct that in the MVP, the signal is only connected once, even if I add three nodes via code. But, in my production/full game code, the signal is added thrice:

```gdscript
# Main scene
for i in range(3):
  _monsters.add_child(Monster.Instantiate())

# Monster container
for child in monsters: print("Connected %s" % monster)

# Logs
Connected Monster:Control<1>
Connected @Control@5:<Control#1234>
Connected @Control@6:<Control#2345>
```

There's something weird going on here.

**Comment by kleonc** [comment]
>But, in my production/full game code, the signal is added thrice:
>
>There's something weird going on here.

I'd say it's unlikely it's caused by a bug in Godot as things like ordering of `_ready` etc. are fundamental and thus bugs involving these would rather be massively reported. And that's not the case. Hence please check out your own logic first.

Anyway, for help with your own project please firstly try using different [community channels (like Discord, Reddit, Q&A, etc.)](https://godotengine.org/community/) as this is a bugtracker, not a helpdesk.

Of course if you'd provide a an actual bug reproduction this issue could be reopened.

**Comment by nightblade9** [comment]
I honesty thought it's a Godot bug, based on how it works for nodes in scene vs. nodes added in code. You're right that this problem would be massively reported if it was legitimate. 

Sorry to waste your time.

**Comment by nightblade9** [comment]
Got around to verifying this today. It also plays well with other tiles (non-autotiles) with minimal effort. Thank you again for making this!
